File: animations/waves.py
import math

from core import Animation, BLACK_BG, RESET, density_char, gray_fg, smoothstep


WAVE_CHAR_THRESHOLDS = [(0.16, " "), (0.24, "."), (0.33, ":"), (0.44, "-"), (0.55, "="), (0.68, "+"), (0.82, "*"), (0.94, "#"), (1.01, "%")]


def wave_char(level):
  return density_char(level, WAVE_CHAR_THRESHOLDS)


class WavesAnimation(Animation):
  def render(self, width, height, elapsed, phase, options):
    density = max(0.45, options.scale)
    t = elapsed
    rows = []
    for row in range(height):
      line = [BLACK_BG]
      last_color = None
      y = row / max(1, height - 1)
      for col in range(width):
        x = col / max(1, width - 1)
        shore = 0.70
        shore += 0.10 * math.sin(2.0 * math.pi * x + 0.20 * math.sin(t * 0.19))
        shore += 0.055 * math.sin(5.4 * math.pi * x - t * 0.11)
        shore += 0.025 * math.sin(12.0 * math.pi * x + 0.13 * t)
        water = 1.0 - smoothstep(shore - 0.045, shore + 0.020, y)
        beach = 1.0 - water
        depth = max(0.0, shore - y)
        shallow = 1.0 - smoothstep(0.02, 0.48, depth)
        shoal_gain = 0.28 + 0.90 * shallow
        sandbar = 0.05 * math.sin(3.0 * math.pi * x - 0.23 * t)
        sandbar += 0.025 * math.sin(9.0 * math.pi * x + 0.31 * t)
        refracted_depth = max(0.0, depth + sandbar * shallow)
        packet_a = 0.62 + 0.38 * math.sin(1.4 * x + 0.31 * t)
        packet_b = 0.64 + 0.36 * math.sin(3.6 * x - 0.17 * t + 1.2)
        approach_a = (18.0 * density) * refracted_depth
        approach_a += 1.8 * math.sin(2.4 * math.pi * x + 0.16 * t)
        approach_b = (11.0 * density) * (refracted_depth + 0.06 * math.sin(5.0 * x - 0.2 * t))
        crest_a = smoothstep(0.78, 0.985, math.sin(approach_a - 2.7 * t))
        crest_b = smoothstep(0.80, 0.990, math.sin(approach_b - 1.6 * t + 1.4))
        incoming = water * shoal_gain * (0.54 * packet_a * crest_a + 0.34 * packet_b * crest_b)
        back_axis = depth + 0.10 * math.sin(4.0 * math.pi * x + 0.2 * t)
        backwash = smoothstep(0.80, 0.99, math.sin((13.0 * density) * back_axis + 1.85 * t))
        backwash *= (0.20 + 0.80 * shallow) * (0.65 + 0.35 * math.sin(2.0 * x - 0.41 * t))
        breaker = math.exp(-42.0 * depth)
        foam_spread = smoothstep(0.18, 0.02, depth)
        shore_foam = water * breaker * (0.35 + 0.65 * max(crest_a, crest_b))
        wash_up = beach * math.exp(-28.0 * max(0.0, y - shore))
        wash_up *= 0.35 + 0.65 * smoothstep(0.25, 0.95, math.sin(8.0 * (y - shore) - 1.35 * t))
        ripples = 0.06 * water * (1.0 - shallow) * abs(math.sin(38.0 * depth + 6.0 * x - 1.1 * t))
        wet_sand = beach * (0.10 + 0.16 * math.exp(-16.0 * max(0.0, y - shore)))
        texture = incoming + 0.42 * backwash + 0.35 * foam_spread * max(crest_a, crest_b)
        texture += shore_foam + wash_up + ripples + wet_sand
        texture = min(1.0, texture)
        vignette_x = abs(x - 0.5) * 2.0
        vignette_y = abs(y - 0.5) * 2.0
        vignette = max(0.0, 1.0 - 0.18 * vignette_x * vignette_x - 0.10 * vignette_y * vignette_y)
        level = max(0.0, min(1.0, texture * vignette * options.contrast))
        char = wave_char(level)
        color = gray_fg(level)
        if last_color != color:
          line.append(f"\x1b[38;5;{color}m")
          last_color = color
        line.append(char)
      line.append(RESET)
      rows.append("".join(line))
    return "\n".join(rows)

File: src/animations/waves.rs
use std::f64::consts::PI;
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, smoothstep, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ocean" };
const TH: &[(f64, char)] = &[
  (0.16, ' '), (0.24, '.'), (0.33, ':'), (0.44, '-'), (0.55, '='),
  (0.68, '+'), (0.82, '*'), (0.94, '#'), (1.01, '%'),
];

pub struct Waves;

impl Animation for Waves {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width; let h = ctx.height;
    let mut grid = vec![0.0_f64; w * h];
    let density = ctx.options.scale.max(0.45);
    let t = ctx.elapsed;
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    let contrast = ctx.options.contrast;
    for row in 0..h {
      let y = row as f64 / dh;
      let vy = (y - 0.5).abs() * 2.0;
      let base = row * w;
      for col in 0..w {
        let x = col as f64 / dw;
        let mut shore = 0.70;
        shore += 0.10 * (2.0 * PI * x + 0.20 * (t * 0.19).sin()).sin();
        shore += 0.055 * (5.4 * PI * x - t * 0.11).sin();
        shore += 0.025 * (12.0 * PI * x + 0.13 * t).sin();
        let water = 1.0 - smoothstep(shore - 0.045, shore + 0.020, y);
        let beach = 1.0 - water;
        let depth = (shore - y).max(0.0);
        let shallow = 1.0 - smoothstep(0.02, 0.48, depth);
        let shoal_gain = 0.28 + 0.90 * shallow;
        let mut sandbar = 0.05 * (3.0 * PI * x - 0.23 * t).sin();
        sandbar += 0.025 * (9.0 * PI * x + 0.31 * t).sin();
        let refr = (depth + sandbar * shallow).max(0.0);
        let pa = 0.62 + 0.38 * (1.4 * x + 0.31 * t).sin();
        let pb = 0.64 + 0.36 * (3.6 * x - 0.17 * t + 1.2).sin();
        let aa = (18.0 * density) * refr + 1.8 * (2.4 * PI * x + 0.16 * t).sin();
        let ab = (11.0 * density) * (refr + 0.06 * (5.0 * x - 0.2 * t).sin());
        let ca = smoothstep(0.78, 0.985, (aa - 2.7 * t).sin());
        let cb = smoothstep(0.80, 0.990, (ab - 1.6 * t + 1.4).sin());
        let incoming = water * shoal_gain * (0.54 * pa * ca + 0.34 * pb * cb);
        let back_axis = depth + 0.10 * (4.0 * PI * x + 0.2 * t).sin();
        let mut backwash = smoothstep(0.80, 0.99, ((13.0 * density) * back_axis + 1.85 * t).sin());
        backwash *= (0.20 + 0.80 * shallow) * (0.65 + 0.35 * (2.0 * x - 0.41 * t).sin());
        let breaker = (-42.0 * depth).exp();
        let foam_spread = smoothstep(0.18, 0.02, depth);
        let shore_foam = water * breaker * (0.35 + 0.65 * ca.max(cb));
        let mut wash_up = beach * (-28.0 * (y - shore).max(0.0)).exp();
        wash_up *= 0.35 + 0.65 * smoothstep(0.25, 0.95, (8.0 * (y - shore) - 1.35 * t).sin());
        let ripples = 0.06 * water * (1.0 - shallow) * (38.0 * depth + 6.0 * x - 1.1 * t).sin().abs();
        let wet_sand = beach * (0.10 + 0.16 * (-16.0 * (y - shore).max(0.0)).exp());
        let mut texture = incoming + 0.42 * backwash + 0.35 * foam_spread * ca.max(cb);
        texture += shore_foam + wash_up + ripples + wet_sand;
        texture = texture.min(1.0);
        let vx = (x - 0.5).abs() * 2.0;
        let vignette = (1.0 - 0.18 * vx * vx - 0.10 * vy * vy).max(0.0);
        grid[base + col] = clamp(texture * vignette * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
